src/cli.ts

```diff
@@ -151,74 +151,73 @@ function parseArguments(args: Array<string>): CLIOptions {
-      else if (info.isDirectory()) {
-        processDirectory(path);
+  for (const path of paths) {
+    await processPath(path);
+  }
+
+  if (callback) {
+    callback(errors);
+  }
+
+  async function processPath(path: string): Promise<void> {
+    try {
+      let info = await stat(path);
+
+      if (info.isDirectory()) {
+        await processDirectory(path);
       } else {
-        processFile(path);
+        // Process file runs async while we calculate the rest of the files
+        await processFile(path);
       }
-    });
+    } catch (err) {
+      errors.push(err);
+    }
   }
 
-  function processDirectory(path: string): void {
-    readdir(path, (err, children) => {
-      if (err) { errors.push(err); }
-      else {
-        pending.unshift(
-          ...children
-            .filter(child => {
-              if (options.modernizeJS) {
-                return child.endsWith('.js');
-              } else {
-                return child.endsWith('.coffee') ||
-                  child.endsWith('.litcoffee') ||
-                  child.endsWith('.coffee.md');
-              }
-            })
-            .map(child => join(path, child))
-        );
+  async function processDirectoryChildren(children: Array<string>, path: string): Promise<void> {
+    for (let child of children) {
+      try {
+        let childPath = join(path, child);
+        let info = await stat(childPath);
+
+        if (info.isDirectory()) {
+          await processDirectory(childPath);
+        } else if (fileFilter.test(child)) {
+          await processFile(childPath);
+        }
+      } catch (err) {
+        errors.push(err);
       }
-      processNext();
-    });
+    }
   }
 
-  function processFile(path: string): void {
+  async function processDirectory(path: string): Promise<void> {
+    try {
+      let children = await readdir(path);
+      await processDirectoryChildren(children, path);
+    } catch (err) {
+      errors.push(err);
+    }
+  }
+
+  async function processFile(path: string): Promise<void> {
     let extension = path.endsWith('.coffee.md') ? '.coffee.md' : extname(path);
     let outputPath = join(dirname(path), basename(path, extension)) + '.js';
+
     console.log(`${path} → ${outputPath}`);
-    readFile(path, 'utf8', (err, data) => {
-      if (err) {
-        errors.push(err);
-        processNext();
-        return;
-      }
-      let resultCode = runWithCode(path, data, options);
-      writeFile(outputPath, resultCode, err => {
-        if (err) { errors.push(err); }
-        processNext();
-      });
-    });
-  }
 
-  function processNext(): void {
-    if (pending.length > 0) {
-      let nextPath = pending.shift();
-      if (!nextPath) {
-        throw new Error('Expected a next path.');
-      }
-      processPath(nextPath);
-    } else if (callback) {
-      callback(errors);
+    try {
+      let data = await readFile(path, 'utf8');
+      let resultCode = runWithCode(path, data, options);
```

Comment:
Putting this in a try/catch actually causes decaffeinate to swallow all unexpected errors now. If an error comes up that's not a PatchError, it throws here, gets added to `errors`, and then gets ignored since `errors` and the `callback` are never used.